crates/containerd-shim-wasm/src/sandbox/containerd.rs

```diff
@@ -0,0 +1,154 @@
+                .map_ok(|msg| msg.data)
+                .try_concat()
+                .await
+                .map_err(|err| ShimError::Others(err.to_string()))
+        })
+    }
+
+    pub fn get_image_content_sha(&self, image_name: impl ToString) -> Result<String> {
+        self.rt.block_on(async {
+            let name = image_name.to_string();
+            let req = GetImageRequest { name };
+            let req = with_namespace!(req, self.namespace);
+            let digest = ImagesClient::new(self.inner.clone())
+                .get(req)
+                .await
+                .map_err(|err| ShimError::Others(err.to_string()))?
+                .into_inner()
+                .image
+                .ok_or_else(|| {
+                    ShimError::Others(format!(
+                        "failed to get image content sha for image {}",
+                        image_name.to_string()
+                    ))
+                })?
+                .target
+                .ok_or_else(|| {
+                    ShimError::Others(format!(
+                        "failed to get image content sha for image {}",
+                        image_name.to_string()
+                    ))
+                })?
+                .digest;
+            Ok(digest)
+        })
+    }
+
+    pub fn get_image(&self, container_name: impl ToString) -> Result<String> {
+        self.rt.block_on(async {
+            let id = container_name.to_string();
+            let req = GetContainerRequest { id };
+            let req = with_namespace!(req, self.namespace);
+            let image = ContainersClient::new(self.inner.clone())
+                .get(req)
+                .await
+                .map_err(|err| ShimError::Others(err.to_string()))?
+                .into_inner()
+                .container
+                .ok_or_else(|| {
+                    ShimError::Others(format!(
+                        "failed to get image for container {}",
+                        container_name.to_string()
+                    ))
+                })?
+                .image;
+            Ok(image)
+        })
+    }
+
+    // load module will query the containerd store to find an image that has an OS of type 'wasm'
+    // If found it continues to parse the manifest and return the layers that contains the WASM modules
+    // and possibly other configuration layers.
+    pub fn load_modules(&self, containerd_id: impl ToString) -> Result<Vec<oci::WasmLayer>> {
+        let image_name = self.get_image(containerd_id.to_string())?;
+        let digest = self.get_image_content_sha(image_name)?;
+        let manifest = self.read_content(digest)?;
+        let manifest = manifest.as_slice();
+        let manifest = ImageManifest::from_reader(manifest)?;
+
+        let arch = manifest
+            .config()
+            .platform()
+            .as_ref()
+            .ok_or_else(|| ShimError::Others("failed to extract platform".to_string()))?
+            .architecture();
```

Comment:
It would be great to expose the platform in the `WasiContext`.
We could use the os part (`wasip1`/`wasip2`) to choose between modules or components, and `os_features` might be useful to know what components we need.

Also, that could also be used to choose between wasm container vs native container, right?

Reply:
Like this idea of passing OS and Arch to the wasicontext.  For features, my understanding is we have some flexibility and we should follow up on how we decide to hand those.

Reply:
To be clear, I'm not implying we should start using `os_features` right now.
The details on that would need wider concensus and review.
But if we expose the `Platform` object (instead of just the OS / Arch strings), we open the door to that option.

Reply:
agreed!

Reply:
I've passed this on to the WasiContext, but not sure I like it the current way I did it.  input welcome